src/components/Table/GridTable.tsx

```diff
@@ -270,141 +259,73 @@ export function GridTable<R extends Kinded, X extends Only<GridTableXss, X> = an
-        if (row.kind === "header") {
-          headerRows.push([row, makeRowComponent(row, level)]);
-        } else if (row.kind === "totals") {
-          totalsRows.push([row, makeRowComponent(row, level)]);
-        } else if (row.kind === "expandableHeader") {
-          expandableHeaderRows.push([row, makeRowComponent(row, level)]);
-        } else {
-          visibleDataRows.push([row, makeRowComponent(row, level)]);
-          // tuples has already been client-side filtered, so just check collapsed
-          if (children.length && !collapsedIds.includes(row.id)) {
-            visitRows(children, level + 1);
-          }
-        }
-      });
-    }
+    // Once our header rows are created we can organize them in expected order.
+    const tableHeadRows = expandableHeaderRows.concat(headerRows).concat(totalsRows);
 
-    // Call `visitRows` with our post-filtered list
-    const [filteredRowIds, filteredRows] = filterRows(api, columns, maybeSorted, filter);
-    visitRows(filteredRows, 0);
-
-    // Check for any selected rows that are not displayed in the table because they don't
-    // match the current filter, or are no longer part of the `rows` prop. We persist these
-    // selected rows and hoist them to the top of the table.
-    if (sortedKeptSelections.length) {
-      keptSelectedRows.push([keptGroupRow as GridDataRow<R>, makeRowComponent(keptGroupRow as GridDataRow<R>, 1)]);
-      if (!collapsedIds.includes(KEPT_GROUP)) {
-        keptSelectedRows.push(
-          ...sortedKeptSelections.map((row, idx) => {
-            const isLast = idx === sortedKeptSelections.length - 1;
-            return [row, makeRowComponent(row, 1, true, isLast)] as RowTuple<R>;
-          }),
-        );
-      }
+    const tooManyClientSideRows = !!filterMaxRows && visibleDataRows.length > filterMaxRows;
+    if (tooManyClientSideRows) {
+      visibleDataRows = visibleDataRows.slice(0, filterMaxRows + keptSelectedRows.length);
     }
 
-    return [headerRows, visibleDataRows, totalsRows, expandableHeaderRows, keptSelectedRows, filteredRowIds];
-  }, [
-    as,
-    api,
-    filter,
-    maybeSorted,
-    columns,
-    style,
-    rowStyles,
-    maybeStyle,
-    columnSizes,
-    collapsedIds,
-    getCount,
-    keptGroupRow,
-    sortedKeptSelections,
-  ]);
-
-  // Once our header rows are created we can organize them in expected order.
-  const tableHeadRows = expandableHeaderRows.concat(headerRows).concat(totalsRows);
-
-  const tooManyClientSideRows = filterMaxRows && visibleDataRows.length > filterMaxRows;
-  if (tooManyClientSideRows) {
-    visibleDataRows = visibleDataRows.slice(0, filterMaxRows + keptSelectedRows.length);
-  }
-
-  useEffect(() => {
-    tableState.setMatchedRows(filteredRowIds);
-  }, [tableState, filteredRowIds]);
+    return [tableHeadRows, visibleDataRows, keptSelectedRows, tooManyClientSideRows];
+  }, [as, api, style, rowStyles, maybeStyle, columnSizes, getCount, filterMaxRows]);
```

Comment:
**thought:** I wonder if we can / should do anything more to ensure some of these are stable properties, like `style`, `rowStyles` and `maybeStyle`. I think `style` gets memoized, but `maybeStyle` and `rowStyles` are up to the implementer to do so. 
I think for `maybeStyle` we just need to determine a couple boolean attributes from it. So maybe we should determine those outside of this useComputed and pass those in as primitives.
I think then we only need to worry about `rowStyles`, which I believe I've seen users already are memoizing before passing to GridTable.

Reply:
Agreed! I've been meaning to play with a `Stable<...>` branded type for forever...will nudge it up on the todo list, and see if anything interesting comes of it.